### PR TITLE
adds TabularMSA.conservation

### PR DESCRIPTION
**Pull request by gregcaporaso**
Not ready for review, just letting the tests run.

### CODE REVIEW

File: skbio/alignment/_tabular_msa.py [hunk 1]
```diff
@@ -741,6 +741,61 @@ def consensus(self):
...
+            If a degenerate character is encountered at a position and
+            ``nan_on_degenerate`` is ``False``.
+
+        """
+        if self.shape[1] == 0:
+            # check empty alignment here b/c looking up character sets
+            # fails otherwise
+            return np.array([])
+
+        result = []
```
Comment (wasade) [inline skbio/alignment/_tabular_msa.py]: Possible to preallocate? The shape of the result is known here.

File: skbio/alignment/_tabular_msa.py [hunk 2]
```diff
@@ -741,6 +741,61 @@ def consensus(self):
...
 
     @experimental(as_of='0.4.0-dev')
+    def conservation(self, metric, nan_on_degenerate=False, nan_on_gap=True):
+        """Apply metric to compute the position conservation for the alignment
+
+        Parameters
+        ----------
+        metric : callable
+            Metric that should be applied for computing conservation. If
+            callable, must take a string representing the characters at single
```
Comment (wasade) [inline skbio/alignment/_tabular_msa.py]: Why `str` and not the native type?

File: skbio/alignment/_tabular_msa.py [hunk 3]
```diff
@@ -741,6 +741,61 @@ def consensus(self):
...
+            position in the alignment, and the TabularMSA object and return a
+            single value.
```
Comment (wasade) [inline skbio/alignment/_tabular_msa.py]: The code suggests this must be a `float` given the use of `nan`

**Comment by jairideout** [comment]
@wasade FYI this is not ready for review, @gregcaporaso is currently making significant API and implementation changes.

File: skbio/alignment/_tabular_msa.py [hunk 4]
```diff
@@ -741,6 +741,61 @@ def consensus(self):
...
+        nan_on_degenerate : bool, optional
+            If ``True``, positions in the resulting array that correspond to
+            alignment positions containing degenerate characters will have
```
Comment (wasade) [inline skbio/alignment/_tabular_msa.py]: containing -> that contain

File: skbio/alignment/_tabular_msa.py [hunk 5]
```diff
@@ -741,6 +741,61 @@ def consensus(self):
...
+            alignment positions with gaps will have np.nan as their value.
+
+        Returns
+        -------
+        np.array
+            Value of applying ``metric`` to each position in the alignment.
+
+        Raises
+        ------
+        KeyError
```
Comment (wasade) [inline skbio/alignment/_tabular_msa.py]: I think a `ValueError` would be more appropriate as this is thrown when a value contained in the position is not supported under this condition.

**Comment by wasade** [comment]
@jairideout k

File: skbio/alignment/_tabular_msa.py [hunk 6]
```diff
@@ -837,6 +838,148 @@ def consensus(self):
...
+    def conservation(self, metric='inverse_shannon_uncertainty',
+                     nan_on_degenerate=False, gap_mode='nan'):
+        """Apply metric to compute conservation for all alignment positions
+
+        Parameters
+        ----------
+        metric : str, optional, {'inverse_shannon_uncertainty'}
+            Metric that should be applied for computing conservation. Resulting
+            values should be larger when a position is more conserved.
+        nan_on_degenerate : bool, optional
```
Comment (gregcaporaso) [inline skbio/alignment/_tabular_msa.py]: Update this to be ``degenerate_mode`` currently supporting only ``nan`` and ``error``.

File: skbio/alignment/_tabular_msa.py [hunk 7]
```diff
@@ -837,6 +838,148 @@ def consensus(self):
...
+        return f
+
+    @experimental(as_of='0.4.0-dev')
+    def conservation(self, metric='inverse_shannon_uncertainty',
+                     nan_on_degenerate=False, gap_mode='nan'):
+        """Apply metric to compute conservation for all alignment positions
+
+        Parameters
+        ----------
+        metric : str, optional, {'inverse_shannon_uncertainty'}
```
Comment (gregcaporaso) [inline skbio/alignment/_tabular_msa.py]: ``optional`` should come last

File: skbio/alignment/_tabular_msa.py [hunk 8]
```diff
@@ -837,6 +838,148 @@ def consensus(self):
...
+            If an unknown ``metric`` is provided.
+        ValueError
+            If any degenerate characters are present in the alignment when
+            ``nan_on_degenerate`` is ``False``.
+        ValueError
+            If any gaps are present in the alignment when ``gap_mode`` is
+            ``"error"``.
+        ValueError
+            If an unknown ``gap_mode`` is provided when there are gaps in the
+            alignment.
```
Comment (gregcaporaso) [inline skbio/alignment/_tabular_msa.py]: Change this so it always errors on unknown gap mode.

**Comment by ebolyen** [comment]
fixes #1087

**Pull request merged**